### PR TITLE
feat: finish multipart io

### PR DESCRIPTION
**Pull request by bojiang**
## Checklist:

- [ ] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/_internal/io_descriptors/multipart.py [hunk 1]
```diff
@@ -27,7 +29,7 @@ def predict(image: "PIL.Image.Image", annotations: t.Dict[str, t.Any]) -> t.Dict
     curl -i -F image=@test.jpg -F annotations=@test.json localhost:5000/predict
     """
 
-    def __init__(self, **inputs: t.Dict[str, IODescriptor]):
+    def __init__(self, **inputs: IODescriptor):
```
Comment (parano) [inline bentoml/_internal/io_descriptors/multipart.py:32]: do you mean inputs: MultiPartIO here?

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/multipart.py:32]: I think typehint `**` can't take a `t.Dict`

File: bentoml/_internal/io_descriptors/multipart.py [hunk 2]
```diff
@@ -55,44 +59,26 @@ def openapi_responses_schema(self):
...
-        :param obj:
-        :type obj: MultipartIO
-        :rtype: Response
-        """
-        pass
+        for k, i in self._inputs.items():
+            req = reqs[k]
+            v = await i.from_http_request(req)
+            res[k] = v
+        print(res)
```
Comment (parano) [inline bentoml/_internal/io_descriptors/multipart.py:75]: remove?

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/multipart.py:75]: yep, I'm fixing the return type

File: bentoml/_internal/utils/_formparser.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        ori_headers = t.cast(t.Dict[bytes, bytes], ori_headers)
+        ori_headers.update(dict(headers))
+        scope["headers"] = list(ori_headers.items())
+        req = Request(scope)
+        req._body = data
+        reqs[field_name] = req
+    return reqs
+
+
+class _MultiResponse(Response):
```
Comment (aarnphm) [inline bentoml/_internal/utils/_formparser.py:133]: @bojiang

File: bentoml/_internal/io_descriptors/image.py [hunk 4]
```diff
@@ -21,11 +19,10 @@
 
     # NOTE: pillow-simd only benefits users who want to do preprocessing
     # TODO: add options for users to choose between simd and native mode
-    _exc = const.IMPORT_ERROR_MSG.format(
```
Comment (parano) [inline bentoml/_internal/io_descriptors/image.py:24]: 👍 we should probably get rid of the IMPORT_ERROR_MSG constant, an inline string like this makes it more readable

File: bentoml/_internal/io_descriptors/multipart.py (hunk 1 above)
Comment (parano) [reply bentoml/_internal/io_descriptors/multipart.py:34]: looks like `**` just assumes all the args are the same type, so `IODescriptor` here is correct

File: bentoml/_internal/runner/runner.py [hunk 5]
```diff
@@ -281,6 +280,7 @@ def run_batch(self, *args, **kwargs):
         if self._state is RunnerState.INIT:
             self._setup()
         if isinstance(self._runner, Runner):
+            print(args, kwargs)
```
Comment (parano) [inline bentoml/_internal/runner/runner.py:283]: remove?

File: bentoml/_internal/server/service_app.py [hunk 6]
```diff
@@ -289,7 +297,7 @@ async def api_func(
                     )
                 else:
                     response = JSONResponse("", status_code=e.error_code)
-            except Exception:  # pylint: disable=broad-except
+            except _StopIteration:
```
Comment (parano) [inline bentoml/_internal/server/service_app.py:300]: What's the reason for this change?

File: bentoml/_internal/tracing/jaeger.py [hunk 7]
```diff
@@ -2,7 +2,7 @@
 from contextlib import contextmanager
 from contextvars import ContextVar
 
-from bentoml._internal.tracing import Tracer
+from ..tracing import Tracer
```
Comment (parano) [inline bentoml/_internal/tracing/jaeger.py:5]: this can be just `from . import Tracer`?

File: bentoml/_internal/server/service_app.py (hunk 6 above)
Comment (aarnphm) [reply bentoml/_internal/server/service_app.py:300]: remove the pylint error

File: bentoml/_internal/tracing/jaeger.py (hunk 7 above)
Comment (aarnphm) [reply bentoml/_internal/tracing/jaeger.py:5]: will fix it

File: bentoml/_internal/io_descriptors/image.py (hunk 4 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/image.py:24]: we already removed the import_error_msg

File: bentoml/_internal/utils/_formparser.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+from ...exceptions import BentoMLException
+
+_ItemsBody = t.TypeVar(
+    "_ItemsBody",
+    bound=t.List[t.Tuple[str, t.List[t.Tuple[bytes, bytes]], bytes]],
+)
+
+
+class MultiPartParser(_StarletteMultiPartParser):
```
Comment (bojiang) [inline bentoml/_internal/utils/_formparser.py:30]: Relying on external codes is not recommended since they may change without warning. copy/paste is much better.

File: bentoml/_internal/utils/_formparser.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                    #         content_type = content_type.decode('latin-1')
+                    #     )
+                    # else:
+                    #     file = None
+                elif message_type == MultiPartMessage.PART_DATA:
+                    data += message_bytes
+                    # if not file:
+                    #     data += message_bytes
+                    # else:
+                    #     await file.write(message_bytes)
```
Comment (bojiang) [inline bentoml/_internal/utils/_formparser.py]: why does we keep these lines?

File: bentoml/_internal/utils/_formparser.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        req = Request(scope)
+        req._body = data
+        reqs[field_name] = req
+    return reqs
+
+
+async def concat_multipart_responses(
+    responses: t.List[t.Tuple[str, t.Union[Response, StreamingResponse]]]
+) -> Response:
+    boundary = binascii.hexlify(os.urandom(16)).decode("ascii")
```
Comment (bojiang) [inline bentoml/_internal/utils/_formparser.py]: i think we can use multipart utils to do this rather than concat manually.

File: bentoml/_internal/utils/_formparser.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
+import binascii
```
Comment (bojiang) [inline bentoml/_internal/utils/_formparser.py:1]: why the renaming?

File: bentoml/_internal/server/service_app.py (hunk 6 above)
Comment (bojiang) [reply bentoml/_internal/server/service_app.py:300]: we want to catch all the exceptions caused by the user function. It's necessary.

Comment (parano) [reply bentoml/_internal/server/service_app.py:300]: @aarnphm let's change this back to `Exception`

File: bentoml/_internal/utils/_formparser.py (hunk 11 above)
Comment (parano) [reply bentoml/_internal/utils/_formparser.py:1]: there's no need to use leading `_` for files like this right? it's already under `_internal`

File: bentoml/_internal/server/service_app.py (hunk 6 above)
Comment (aarnphm) [reply bentoml/_internal/server/service_app.py:300]: ok

File: bentoml/_internal/utils/_formparser.py (hunk 11 above)
Comment (aarnphm) [reply bentoml/_internal/utils/_formparser.py:1]: you are correct

File: bentoml/_internal/io_descriptors/image.py [hunk 12]
```diff
@@ -1,11 +1,11 @@
 import io
+import mimetypes
 import typing as t
 
+import imageio
```
Comment (parano) [inline bentoml/_internal/io_descriptors/image.py:5]: `imageio` is no longer a dependency, is it possible to remove it? ideally we want to use only `pillow`, which is imported with lazyloader below. cc @aarnphm @bojiang

Comment (parano) [reply bentoml/_internal/io_descriptors/image.py:5]: this is breaking the documentation checker rn

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/image.py:5]: I will update it to Pillow

Comment (bojiang) [reply bentoml/_internal/io_descriptors/image.py:5]: oh never expected that. yeah these commits is for quick testing, not well formated for dependencies
